web/src/compose.js

```diff
@@ -169,7 +178,137 @@ export function send_message_success(request, data) {
+ */
+function check_need_to_split_long_messages() {
+    if (!is_split_long_messages_enabled()) {
+        return false;
+    }
+    const content = compose_state.message_content();
+    return content.includes(SPLIT_DELIMITER);
+}
+
+export function is_split_long_messages_enabled() {
+    return localStorage.getItem("split_long_messages") === "true";
+}
+
+function show_split_long_messages_info_banner() {
+    if (split_long_messages_banner_shown) {
+        return;
+    }
+    const context = {
+        banner_type: compose_banner.INFO,
+        banner_text: $t({
+            defaultMessage: `Your message will be split into different parts.`,
+        }),
+        button_text: $t({
+            defaultMessage: `Send as a single message`,
+        }),
+        classname: compose_banner.CLASSNAMES.split_long_messages,
+    };
+    const new_row_html = render_compose_banner(context);
+    compose_banner.append_compose_banner_to_banner_list($(new_row_html), $("#compose_banners"));
+    split_long_messages_banner_shown = true;
+}
+
+function clear_split_long_messages_info_banner() {
+    if (!split_long_messages_banner_shown) {
+        return;
+    }
+    $(`#compose_banners .${CSS.escape(compose_banner.CLASSNAMES.split_long_messages)}`).remove();
+    split_long_messages_banner_shown = false;
+}
+
+export function toggle_split_long_messages() {
+    let state = is_split_long_messages_enabled();
+    if (state) {
+        localStorage.setItem("split_long_messages", "false");
+    } else {
+        localStorage.setItem("split_long_messages", "true");
+    }
+    state = !state;
+
+    // preview area and compose banners should be updated with the new setting
+    if (is_preview_area_showing()) {
+        // update the preview to show/hide horizontal lines showing where the content will be split
+        show_preview_area();
+    }
+    update_split_long_messages_info_banner();
+    const $toggle_button = $(".split_long_messages_toggle");
+    if ($toggle_button.length === 0) {
+        return;
+    }
+    $toggle_button.text(
+        $t(
+            {defaultMessage: "{split_long_messages_toggle_button_text}"},
+            {
+                split_long_messages_toggle_button_text: state
+                    ? "Send as a single message"
+                    : "Send as multiple messages",
+            },
+        ),
+    );
+}
+
+/**
+ * @param {string} content the content to be split
+ * @returns a list of non-empty message parts, split by the delimiter
+ */
+export function message_content_split_parts(content = compose_state.message_content()) {
+    return content.split(SPLIT_DELIMITER).filter((k) => k.length > 0);
+}
+
+export function send_message() {
+    if (!check_need_to_split_long_messages()) {
+        send_individual_message();
+        return;
+    }
+    const split_parts = message_content_split_parts();
+    let i = -1;
+    // start the message sending chain
+    send_next_message(true);
```

Comment:
I don't understand what you're doing here -- is this a loop or recursion? I think recursion might be more appropriate in any case.

Reply:
It is indeed recursion. I have created a callback `send_next_message` which is invoked after one part of the message is sent. I am using the variable `i` to keep track of the message parts which are going to be sent. The variable `i` is captured by the callback.

The callback is called with true or false depending on the success of the message part just sent, and the appropriate action is then taken by the callback. 

I tried to add on to the existing code without making too many modifications.